### ISSUE TITLE
Question: Can this be used for UDP sockets?

### ISSUE THREAD

**Issue by wajahat-abbas** [ARMmbed/mbed-os-example-tls-socket#28]
I would like to use UDP with TLS (DTLS). Can this example be used for UDP or is there any other such example for DTLS?

**Comment by SeppoTakalo** [comment]
API for `DTLSSocket` is pretty much the same as for `TLSSocket`, therefore separate example is not provided.

See the API from:
https://os.mbed.com/docs/mbed-os/v5.13/apis/dtlssocket.html

Also, there are very little public DTLS servers that could be used in such simple fashion.
You could run your own DTLS server and easily modify this example.

**Comment by wajahat-abbas** [comment]
Great! Thank you.

**Comment by wajahat-abbas** [comment]
@SeppoTakalo Another quick question: How can i setup mbedTLS for PSK? The method `set_client_cert_key` expects certificate and key where if i am not wrong, for PSK, certificate is not needed. Can you point me to APIs for PSK?

**Comment by ciarmcom** [comment]
*Internal Jira reference: https://jira.arm.com/browse/MBOCUSTRIA-1462*

**Comment by wajahat-abbas** [comment]
Also i modified the example with DTLS but getting hardfault inside socket.connect function.

**Comment by SeppoTakalo** [comment]
For configuring the PSK, or any other mode, you need to fetch the internal Mbed TLS object pointers from `TLSSocket` object, then use Mbed TLS API for configuration.
See: https://os.mbed.com/docs/mbed-os/v5.13/reference/secure-socket.html and scroll down for "Advanced usage"

**Comment by SeppoTakalo** [comment]
For hard fault, create separate issue ticket, where you submit your code changes, and back-trace with debugger.

**Comment by wajahat-abbas** [comment]
OK thank you and i suppose that is not related with this repository either so closing this.

**Comment by syhoangmai** [comment]
Hi,

Just want to share an experience of mine with Mbed DTLS socket. Previously I asked the same question about psk but received no solution so I had to improvise a bit. The function that can be used to set psk is mbedtls_ssl_conf_psk(). I have not seen a hard fault when using this, though (mbed 5.12). I also managed to authenticate an mbed device with Leshan server via psk dtls, so the socket psk works. Hope this helps.

By the way, could this functionality be added into the official Mbed (D)TLS socket as it is very short? I think even when certificate used, the secure socket will always comeback to pre-shared key anyway.
